pkg/sns/topic_test.go

```diff
@@ -43,133 +36,179 @@ func TestTopic_Publish(t *testing.T) {
 		TopicId: "topic",
 	}
 
-	topic := sns.NewTopicWithInterfaces(logger, client, exec, s)
-	err := topic.Publish(context.Background(), aws.String("test"))
-
-	assert.NoError(t, err)
-
-	client.AssertExpectations(t)
+	s.client = new(snsMocks.Client)
+	s.executor = gosoAws.NewTestableExecutor([]gosoAws.TestExecution{})
+	s.topic = sns.NewTopicWithInterfaces(logger, s.client, s.executor, settings)
 }
 
-func TestTopic_PublishError(t *testing.T) {
-	logger := monMocks.NewLoggerMockedAll()
+func (s *TopicTestSuite) addExecution(output interface{}, err error) {
+	s.executor.ExpectExecution(gosoAws.TestExecution{
+		Output: output,
+		Err:    err,
+	})
+}
 
+func (s *TopicTestSuite) TestPublish() {
 	input := &awsSns.PublishInput{
-		TopicArn: aws.String("arn"),
-		Message:  aws.String("test"),
+		TopicArn:          aws.String("topicArn"),
+		Message:           aws.String("test"),
+		MessageAttributes: map[string]*awsSns.MessageAttributeValue{},
 	}
+	s.client.On("PublishRequest", input).Return(nil, nil)
+	s.addExecution(nil, nil)
 
-	exec := gosoAws.NewTestableExecutor([]gosoAws.TestExecution{{
-		Output: nil,
-		Err:    errors.New("error"),
-	}})
+	err := s.topic.Publish(context.Background(), aws.String("test"), map[string]interface{}{})
+	s.NoError(err)
 
-	client := new(snsMocks.Client)
-	client.On("PublishRequest", input).Return(nil, nil)
+	s.True(s.executor.HasExecutedAll())
```

Comment:
If this is the only use of `HasExecutedAll`, maybe it should be replaced by some `AssertExpectations` function

Reply:
yeah, that was my first idea, too, but then you have a dependency to the testing package in `cloud/aws`..  
that doesn't felt nice..

Reply:
Does it make sense to move the TestableExecutor to a test package? As it's only supposed to be used in tests? There you would be able to depend on the testing package, right?

Or replace the executor entirely by a mock?
I don't see the benefit of having a stub here, especially as you're reimplementing the expectations